pkg/controller/certificates/sync.go

```diff
@@ -79,6 +79,12 @@ func (c *Controller) Sync(ctx context.Context, crt *v1alpha1.Certificate) (err e
 		}
 	}()
 
+	// inject default key algorithm and size
+	if crtCopy.Spec.KeyAlgorithm == v1alpha1.KeyAlgorithm("") && crtCopy.Spec.KeySize == 0 {
+		crtCopy.Spec.KeyAlgorithm = c.CertificateOptions.DefaultKeyAlgorithm
+		crtCopy.Spec.KeySize = c.CertificateOptions.DefaultKeySize
```

Comment:
We should be cautious here actually updating the certificate's spec to reflect the new/defaulted values. Controller's should only really modify the `status` subresource of the resource they are managing, and actual defaulting should instead happen as part of a mutating admission hook.

Could you refactor this PR to not require modifying the Certificate itself? Probably by creating a keyAlgorithm and keySize var here, and updating all usages of these fields to use the defaulted in-memory values?

Reply:
This will probably require you to discover usages of both these fields across the whole codebase to do properly.

Hopefully the majority of the places that use this field are in this `pkg/util/pki` package.

It may be best to create a new function that handles the defaulting logic, like we do for the RenewBefore/Duration fields. In future these methods will probably move into `pkg/api/util`, but that is something for another day 😄

Reply:
I decided I won't be doing global defaults, just defaults at the ingress shim level. I may try to raise a PR for global defaults at a later time.